contracts/Lottery.sol

```diff
@@ -19,72 +20,103 @@ contract Lottery {
+        emit NewUser(msg.sender);
+
     }
     
-    function startLottery(uint32 _ticketPrice) public{
+    function startLottery(uint _ticketPrice) public{
         require (owner == msg.sender, "Only owner can open lottery process.");
-        state = LotteryState.Started;
         ticketPrice = _ticketPrice;
+        state = LotteryState.Started;
     }
-
-    function stopLottery() public{
-        require (owner == msg.sender, "Only owner can open lottery process.");
-        state = LotteryState.Finished;
+    
+    function terminateLottery() public {
+        require (owner == msg.sender, "Only owner can close lottery.");
+        require (state == LotteryState.Started, "To cancel lottery contract, it should be in Started state");
+        state = LotteryState.Terminated;
+        emit LotteryStatusUpdateWithMessage(msg.sender, getLotteryState(), "Lottery has been terminated, you will receive refund shortly.");
+        
+        //Refund process for every player
+        for (uint counter = 0; counter < players.length; counter++) {
+            Player memory player = players[counter];
+            LotteryTicket[] memory tickets = playerLotteryTicketsMapping[player.playerAddress];
+            uint playerAmount = tickets.length * ticketPrice;
+            if(playerAmount > 0) {
+                player.playerAddress.transfer(playerAmount);
+                emit RefundLottery(player.playerAddress, playerAmount);
+            }
+        }
         
     }
     
-    function buyLotteryTickets(uint32 _ticketNumbers) public payable returns (string memory result) {
-        require(msg.value >= (_ticketNumbers*ticketPrice), "Insufficient balance.");
-        require(state == LotteryState.Started, "Lottery is not yet opened");
+    
+    function buyLotteryTickets(uint count, uint value) public payable {
         require(owner != msg.sender, "Lottery owner/organizer can't buy tickets.");
+        require(state == LotteryState.Started, "Lottery is not yet opened");
         require(playerMap[msg.sender].isExist,"User not registered.");
+        require(msg.sender.balance > value, "Insufficient balance to purchase ticket(s).");
+        //uint memory ticketsPrice = count.mul(ticketPrice);
+        uint  ticketsPrice = count * ticketPrice;
+        require(value == ticketsPrice, "There is difference between actual price and provided price.");
+        require(msg.value == ticketsPrice, "Amount sent for buying ticket(s) are not sufficient");
         
-        for(uint i=0; i<_ticketNumbers; i++) {
-           LotteryTicket memory ticket = LotteryTicket(tickets.length+1,ticketPrice, msg.sender);
-            totalAmount = totalAmount + ticketPrice;
+        for(uint counter=1; counter <= count; counter++) {
+            //LotteryTicket memory ticket = LotteryTicket(tickets.length.add(count),ticketPrice, msg.sender);
+            LotteryTicket memory ticket = LotteryTicket(tickets.length + count,ticketPrice, msg.sender);
```

Comment:
This code looks like it will increment the ticket id by `count` instead of 1 like the previous code. So, if I buy 5 tickets, they would be 5, 10, 15, 20, 25.

This could be reverted to the previous version I think.

Reply:
Corrected

Reply:
This still feels wrong. You are incrementing by count so it will always skip alternate values. For example, the following code snippet:

```javascript
var count = 5;
var tickets = [];
for(counter=1; counter <= count; counter++) {
	var ticket = tickets.length + counter;
	tickets.push(ticket);
}

console.log(tickets)
```

produces the following output:

```javascript 
[1, 3, 5, 7, 9] 
```

Hope you get my point!!